[Style] Convert widows and orphans to strong style types

weinig opened the pull request:
#### cc58756c6140ddc18fa899b1b4049ade7d68f3c2
<pre>
[Style] Convert widows and orphans to strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=296767">https://bugs.webkit.org/show_bug.cgi?id=296767</a>

Reviewed by Darin Adler.

Converts the `widows` and `orphans` properties to use strong styles.

The current implementation of `widows` and `orphans` is a bit bizarre.
Both properties have an &quot;is auto&quot; bit in RenderStyle, which gets set
as the initial value, despite neither allowing `auto` as a value at
parse time. This leads to the computed style not being parseable.
This change does not attempt to fix this.

To ensure correct numeric clamping, instead of clamping when setting
on RenderStyle, we utilize the range template parameter and clamp enforce
the clamp when converting from CSSValue, ensuring the value is never
invalid. This allows us to also remove now unnecessary clamping for
the `shape-image-threshold` property.

* Source/WTF/wtf/MathExtras.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericRange.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationPagination.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/break/StyleOrphans.cpp: Added.
* Source/WebCore/style/values/break/StyleOrphans.h: Added.
* Source/WebCore/style/values/break/StyleWidows.cpp: Added.
* Source/WebCore/style/values/break/StyleWidows.h: Added.
* Source/WebCore/style/values/color/StyleOpacity.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:

Canonical link: <a href="https://commits.webkit.org/298403@main">https://commits.webkit.org/298403@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/552d4928422715bd641f45bfeebe74cf3b636419

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115414 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25610 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121503 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65992 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117303 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35767 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43679 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87696 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118362 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28526 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103601 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68088 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27684 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21723 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65156 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/107619 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97915 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124668 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/113953 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31730 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42721 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99789 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/96263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24486 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41491 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19347 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38258 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42233 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47789 "Built successfully") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/139051 "Build was cancelled. Recent messages:Failed to print configuration") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41735 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/139051 "Build was cancelled. Recent messages:Failed to print configuration") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45063 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/MathExtras.h:
```diff
@@ -256,13 +256,12 @@ template<typename TargetType, typename SourceType>
-    if (value < 0)
-        return 0;
-    TargetType max = std::numeric_limits<TargetType>::max();
     if (value >= static_cast<SourceType>(max))
         return max;
+    if (value <= static_cast<SourceType>(min))
+        return min;
```
Why do the max before the min, since the old code did the min first? I suppose it doesn't matter as long as they are ordered. I've always thought that if somehow min and max were out of order we would want the min to be returned. Also, if these are known at compile time, is there some way to make sure they are never out of order?

weinig replied:
I was matching all the other clampTo functions above it that do max than min.

darinadler commented on Source/WebCore/css/values/primitives/CSSPrimitiveNumericRange.h:
```diff
@@ -114,6 +114,56 @@ template<Range range, std::floating_point T> constexpr T clampToRange(T value, T
+        return clampTo<T>(
+            value,
+            range.min,
+            std::min<T>(range.max,  std::numeric_limits<T>::max())
```
```suggestion
            std::min<T>(range.max, std::numeric_limits<T>::max())
```
Also, is there a guarantee this can be done at compile time?

weinig replied:
I wish c++ had some keyword/syntax to say "run this function only at compile time, erroring otherwise" like:

```c++
[[require_consteval]] std::min<T>(range.max, std::numeric_limits<T>::max())
```

but none exists. That said, these are all `constexpr` values so it will happen at compile time.

darinadler commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
```diff
@@ -39,59 +39,59 @@ template<auto R, typename V> struct CSSValueConversion<Integer<R, V>> {
     {
         Ref protectedValue = value;
-        return { protectedValue->resolveAsNumber<V>(builderState.cssToLengthConversionData()) };
+        return { CSS::clampToRange<R, V>(protectedValue->resolveAsNumber<V>(builderState.cssToLengthConversionData())) };
```
Is there a way to make this repeated pattern easier to get right? What if you left out one of these by accident? Is there a test that covers every single one of them?

weinig replied:
I can't think of one off the top of my head, but the intent is that mostly people shouldn't have to remember and all the clamping happens in this file unless you are doing something non-standard.

As for testing, these are kinda belt-and-suspenders at the moment. as in most cases this clamping is happening at parse time or via calc() resolution. They are protecting against the cases where the parser has weird rules (like for opacity values) that aren't encoded in the type system yet. My hope is that these will eventually turned into assertions.

darinadler reviewed (approved):
I suppose it may be a little premature to review this as it seems tests are still failing.

weinig committed:
[Style] Convert widows and orphans to strong style types
https://bugs.webkit.org/show_bug.cgi?id=296767

Reviewed by Darin Adler.

Converts the `widows` and `orphans` properties to use strong styles.

The current implementation of `widows` and `orphans` is a bit bizarre.
Both properties have an "is auto" bit in RenderStyle, which gets set
as the initial value, despite neither allowing `auto` as a value at
parse time. This leads to the computed style not being parseable.
This change does not attempt to fix this.

To ensure correct numeric clamping, instead of clamping when setting
on RenderStyle, we utilize the range template parameter and clamp enforce
the clamp when converting from CSSValue, ensuring the value is never
invalid. This allows us to also remove now unnecessary clamping for
the `shape-image-threshold` property.

* Source/WTF/wtf/MathExtras.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericRange.h:
* Source/WebCore/layout/integration/inline/LayoutIntegrationPagination.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/break/StyleOrphans.cpp: Added.
* Source/WebCore/style/values/break/StyleOrphans.h: Added.
* Source/WebCore/style/values/break/StyleWidows.cpp: Added.
* Source/WebCore/style/values/break/StyleWidows.h: Added.
* Source/WebCore/style/values/color/StyleOpacity.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Blending.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:

Canonical link: https://commits.webkit.org/298403@main
